src/webgpu/shader/validation/expression/call/builtin/bitcast.spec.ts

```diff
@@ -0,0 +1,141 @@
+  // for negative infinity, plus one.
+  // The negative NaN with the highest integer representation is u32.max (!)
+  ...linearRange(kBit.f32.infinity.negative + 1, kBit.u32.max, numNaNs),
+  kBit.f32.infinity.positive,
+  kBit.f32.infinity.negative,
+];
+
+g.test('bad_const_to_f32')
+  .specURL('https://www.w3.org/TR/WGSL/#floating-point-evaluation')
+  .desc(
+    `
+It is a shader-creation error if any const-expression of floating-point type evaluates to NaN or infinity.
+`
+  )
+  .params(u =>
+    u
+      .combine('fromScalarType', ['i32', 'u32'] as const)
+      .combine('vectorize', [...vectorCases()] as const)
+      .combine('bitBadValue', [...f32InfAndNaNInU32] as const)
+  )
+  .fn(t => {
+    // For scalar cases, generate code like:
+    //  const f = bitcast<f32>(i32(u32(0x7f800000)));
+    // For vector cases, generate code where one component is bad. In this case
+    // width=4 and badIndex=2
+    //  const f = bitcast<vec4f>(vec4<32>(0,0,i32(u32(0x7f800000)),0));
+    const width = t.params.vectorize.width;
+    const badIndex = t.params.vectorize.badIndex;
+    const badScalar = `${t.params.fromScalarType}(u32(${t.params.bitBadValue}))`;
+    const destType = width === 1 ? 'f32' : `vec${width}f`;
+    const srcType =
+      width === 1 ? t.params.fromScalarType : `vec${width}<${t.params.fromScalarType}>`;
+    const components = [...Array(width).keys()]
+      .map(i => (i === badIndex ? badScalar : '0'))
+      .join(',');
+    const code = `const f = bitcast<${destType}>(${srcType}(${components}));`;
+    t.expectCompileResult(false, code);
+  });
+
+const f32_matrix_types = [2, 3, 4]
+  .map(i => [2, 3, 4].map(j => `mat${i}x${j}f`))
+  .reduce((a, c) => a.concat(c), []);
+const bool_types = ['bool', ...[2, 3, 4].map(i => `vec${i}<bool>`)];
+
+g.test('bad_type_constructible')
+  .specURL('https://www.w3.org/TR/WGSL/#bitcast-builtin')
+  .desc(
+    `
+Bitcast only applies to concrete numeric scalar or concrete numeric vector.
+Test constructible types.
+`
+  )
+  .params(u =>
+    u
+      .combine('type', [...f32_matrix_types, ...bool_types, 'array<i32,2>', 'S'])
+      .combine('direction', ['to', 'from'])
+  )
+  .fn(t => {
+    const T = t.params.type;
+    const preamble = T === 'S' ? 'struct S { a:i32 } ' : '';
+    // Create a value of type T using zero-construction: T().
+    const srcVal = t.params.direction === 'to' ? '0' : `${T}()`;
+    const destType = t.params.direction === 'to' ? T : 'i32';
+    const code = preamble + `const x = bitcast<${destType}>(${srcVal});`;
+    t.expectCompileResult(false, code);
+  });
+
+g.test('bad_type_nonconstructible')
+  .specURL('https://www.w3.org/TR/WGSL/#bitcast-builtin')
+  .desc(
+    `
+Bitcast only applies to concrete numeric scalar or concrete numeric vector.
+Test non-constructible types.
+`
+  )
+  .params(u => u.combine('var', ['s', 't', 'b', 'p']).combine('direction', ['to', 'from']))
+  .fn(t => {
+    const typeOf: Record<string, string> = {
```

Comment:
nit: I think because the keys are 'string's, the type `Record<string, string>` can be omitted here. (and I think more idiomatic)

Reply:
If I drop the `Record<string,string>` then I get this error:

```
src/webgpu/shader/validation/expression/call/builtin/bitcast.spec.ts:121:52 - error TS7053: Element implicitly has an 'any' type because expression of type 'string' can't be used to index type '{ s: string; t: string; b: string; p: string; }'.
  No index signature with a parameter of type 'string' was found on type '{ s: string; t: string; b: string; p: string; }'.

121     const destType = t.params.direction === 'to' ? typeOf[t.params.var] : 'i32';
                                                       ~~~~~~~~~~~~~~~~~~~~

```

Reply:
Hmmm, that surprises me, but I am fine with leaving as is.